src/decisionengine/framework/engine/SourceWorkers.py

```diff
@@ -0,0 +1,191 @@
+
+        self.logger = structlog.getLogger(LOGGERNAME)
+        self.logger = self.logger.bind(module=__name__.split(".")[-1], source=self.name)
+
+        self.exchange = exchange
+        self.connection = Connection(broker_url)
+
+        # We use a random name to avoid queue collisions when running tests
+        queue_id = str(uuid.uuid4()).upper()
+        self.logger.debug(f"Creating queue {queue_id} with routing key {self.key}")
+        self.queue = Queue(
+            queue_id,
+            exchange=self.exchange,
+            routing_key=self.key,
+            auto_delete=True,
+        )
+        self.use_count = multiprocessing.Value("i", 1)
+        self.schedule = config.get("schedule", _DEFAULT_SCHEDULE)
+
+        self.logger.debug(
+            f"Creating worker: module={self.module} name={self.key} class_name={self.name} parameters={config['parameters']} schedule={self.schedule}"
+        )
+
+    def should_stop(self):
+        with self.use_count.get_lock():
+            return self.use_count.value == 0
+
+    def increment_use_count(self):
+        with self.use_count.get_lock():
+            self.use_count.value += 1
+
+    def decrement_use_count(self):
+        with self.use_count.get_lock():
+            self.use_count.value -= 1
+
+    def take_offline(self):
+        with self.use_count.get_lock():
+            self.use_count.value = 0
+
+    def run(self):
+        """
+        Get the data from source
+        """
+        self.logger.info(f"Starting source loop for {self.key}")
+        SOURCE_ACQUIRE_GAUGE.labels(self.key)
+        with producers[self.connection].acquire(block=True) as producer:
+            # If task manager is in offline state, do not keep executing sources.
+            while not self.should_stop():
+                try:
+                    self.logger.info(f"Source {self.name} calling acquire")
+                    data = self.module_instance.acquire()
+                    Module.verify_products(self.module_instance, data)
+                    self.logger.info(f"Source {self.name} acquire returned")
+                    SOURCE_ACQUIRE_GAUGE.labels(self.name).set_to_current_time()
+                    self.logger.debug(f"Publishing data to queue {self.key} with routing key {self.key}")
+                    producer.publish(
+                        dict(source_module=self.module, class_name=self.name, data=data),
+                        routing_key=self.key,
+                        exchange=self.exchange,
+                        serializer="pickle",
+                        declare=[
+                            self.exchange,
+                            self.queue,
+                        ],
+                    )
+                    self.logger.info(f"Source {self.name} {self.module} finished cycle")
+                except Exception:
+                    self.logger.exception(f"Exception running source {self.name} ")
+                    producer.publish(
```

Comment:
Did we want to re-raise the exception rather than republish or am I missing something here?

Reply:
When an exception occurs while running the source, any downstream channels must be taken offline.  This is achieved by publishing a message where the data is `State.SHUTDOWN` (see next line).  Any listening channels consume that message and enter the shutting-down state.

The source's exception is logged above and the source loop terminates after publishing the `State.SHUTDOWN` message.

Reply:
Ahh, I get it.  Somehow I didn't see `data=State.SHUTDOWN`.....